corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,103 @@ def delete_old_request_logs():
             timing_buckets=_check_repeaters_buckets,
         ):
-            for record in RepeatRecord.objects.iter_partition(
-                    start, partition, CHECK_REPEATERS_PARTITION_COUNT):
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
+
+                if datetime.utcnow() > twentythree_hours_later:
+                    # Break after process_repeater() to avoid a stale lock
+                    break
     finally:
         check_repeater_lock.release()
 
 
+def iter_ready_repeaters():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for repeater in Repeater.objects.all_ready():
+            if not repeater.domain_can_forward:
+                continue
+
+            if rate_limit_repeater(repeater.domain):
+                repeater.rate_limit()
+                continue
+
+            yielded = True
+            yield repeater
+
+        if not yielded:
+            # No repeaters are ready, or they are rate limited, or their
+            # domains can't forward or are paused.
+            return
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_repeater(domain, repeater_id):
+
+    def is_retry(repeat_record):
+        return repeat_record.state == State.Fail
+
+    repeater = Repeater.objects.get(domain=domain, id=repeater_id)
+    repeat_records = repeater.repeat_records_ready[:repeater.num_workers]
+    header_tasks = [
+        retry_process_repeat_record.s(rr.id, rr.domain)
+        if is_retry(rr)
+        else process_repeat_record.s(rr.id, rr.domain)
+        for rr in repeat_records
+    ]
+    chord(header_tasks)(update_repeater.s(repeater.repeater_id))
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def update_repeater(repeat_record_states, repeater_id):
+    """
+    Determines whether the repeater should back off, based on the
+    results of `fire_repeat_record()` tasks.
+    """
+    repeater = Repeater.objects.get(id=repeater_id)
+    if any(s == State.Success for s in repeat_record_states):
+        # At least one repeat record was sent successfully.
+        repeater.reset_backoff()
+    elif all(s in (State.Empty, None) for s in repeat_record_states):
```

Comment:
Would it make sense to document why a record state would be `None` or `Empty` at this point? Like is it either
- there was an error in HQ code (I think that's one possibility if the `except Exception` block was executed below in `_process_repeat_record`) resulting in `None`
- or because the repeat record had an empty payload that caused it not to be sent resulting in `Empty`

Are there other possibilities?